src/compiler/crystal/compiler.cr

```diff
@@ -546,62 +546,103 @@ module Crystal
         return all_reused
       end
 
-      {% if !Crystal::System::Process.class.has_method?("fork") %}
+      {% if !LibC.has_method?("fork") %}
         raise "Cannot fork compiler. `Crystal::System::Process.fork` is not implemented on this system."
       {% elsif flag?(:preview_mt) %}
         raise "Cannot fork compiler in multithread mode"
       {% else %}
-        jobs_count = 0
-        wait_channel = Channel(Array(String)).new(@n_threads)
+        workers = fork_workers do |input, output|
+          while i = input.gets(chomp: true).presence
+            unit = units[i.to_i]
+            unit.compile
+            result = {name: unit.name, reused: unit.reused_previous_compilation?}
+            output.puts result.to_json
+          end
+        end
+
+        overqueue = 1
+        indexes = Atomic(Int32).new(0)
+        channel = Channel(String).new(@n_threads)
+        completed = Channel(Nil).new(@n_threads)
 
-        units.each_slice(Math.max(units.size // @n_threads, 1)) do |slice|
-          jobs_count += 1
+        workers.each do |pid, input, output|
           spawn do
-            # For stats output we want to count how many previous
-            # .o files were reused, mainly to detect performance regressions.
-            # Because we fork, we must communicate using a pipe.
-            reused = [] of String
-            if wants_stats_or_progress
-              pr, pw = IO.pipe
-              spawn do
-                pr.each_line do |line|
-                  unit = JSON.parse(line)
-                  reused << unit["name"].as_s if unit["reused"].as_bool
-                  @progress_tracker.stage_progress += 1
-                end
+            overqueue.times do
+              if (index = indexes.add(1)) < units.size
+                input.puts index
               end
             end
 
-            codegen_process = Crystal::System::Process.fork do
-              pipe_w = pw
-              slice.each do |unit|
-                unit.compile
-                if pipe_w
-                  unit_json = {name: unit.name, reused: unit.reused_previous_compilation?}.to_json
-                  pipe_w.puts unit_json
-                end
+            while (index = indexes.add(1)) < units.size
+              input.puts index
+
+              if response = output.gets(chomp: true)
```

Comment:
**question:** I'm just wondering what happens if the output is nil for whatever reason. I suppose the fiber will continue and eventually signal `completed` which then closes `channel`. And the entire method completes. But I don't see any error happening. There will just be a compiled unit missing? I guess that'll error at the linking stage, but we could detect this early.
I suppose if `output.gets` is *never* empty, that should be an error. We could try building the unit on a different worker, but that would be a bit complicated because we cannot easily unwind the counter. And it might just fail as well.